sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -105,6 +105,66 @@ trait InvokeLike extends Expression with NonSQLExpression {
   }
 }
 
+/**
+ * Common trait for [[DecodeUsingSerializer]] and [[EncodeUsingSerializer]]
+ */
+trait BaseSerializer {
+  /**
+   * If true, Kryo serialization is used, otherwise the Java one is used
+   */
+  val kryo: Boolean
+
+  /**
+   * The serializer instance to be used for serialization/deserialization
+   */
+  lazy val serializerInstance = {
```

Comment:
We can also put this in a companion and call this from whole stage code gen.

Reply:
The major concern I have on this is performance. Referencing an outer object means that is is put into an array and every time it is called we have an array reference call and a cast before its usage. I saw many concerns in previous PRs on codegen about doing these things in order not to affect performances.

Reply:
I was talking about the code below (the comment was somewhat terse). We should definitely put the `serializerInstance` into a the generated classes immutable state. All I am saying is that is that we can put the factory code below in a companion, and use that in both the generated and interpreted code paths. This reduces the amount of code generated, and also makes things easier to test.

Reply:
Oh, sorry, I see now what you mean. Yes, I will do that, thanks.